drivers/gles3/shader_tracker_gles3.cpp

```diff
@@ -0,0 +1,445 @@
+		usage_flag_pointers.push_back(tmp);
+	}
+	result += _strings_to_csv_string(usage_flag_pointers) + "\n";
+
+	Vector<String> write_flag_pointers;
+	for (Map<StringName, bool *>::Element *e = p_actions->write_flag_pointers.front(); e; e = e->next()) {
+		String tmp = vformat("%s=%d", e->key(), (int)*e->value());
+		write_flag_pointers.push_back(tmp);
+	}
+	result += _strings_to_csv_string(write_flag_pointers) + "\n";
+	result += vformat("%d\n", p_conditional_key);
+
+	return result;
+}
+
+Error ShaderPreLoader::load_spatial(const String &p_file_path) {
+	return _load(VS::ShaderMode::SHADER_SPATIAL, p_file_path);
+}
+
+Error ShaderPreLoader::load_canvas(const String &p_file_path) {
+	return _load(VS::ShaderMode::SHADER_CANVAS_ITEM, p_file_path);
+}
+
+Error ShaderPreLoader::load_particle(const String &p_file_path) {
+	return _load(VS::ShaderMode::SHADER_PARTICLES, p_file_path);
+}
+
+bool ShaderPreLoader::compiling = false;
+int ShaderPreLoader::spatial_count = 0;
+int ShaderPreLoader::canvas_count = 0;
+int ShaderPreLoader::particle_count = 0;
+int ShaderPreLoader::progress = 0;
+
+Map<uint32_t, String> ShaderPreLoader::spatial_shader_code = Map<uint32_t, String>();
+Vector<ShaderPreLoader::ShaderPermutation> ShaderPreLoader::spatial_shaders = Vector<ShaderPreLoader::ShaderPermutation>();
+Map<uint32_t, String> ShaderPreLoader::canvas_shader_code = Map<uint32_t, String>();
+Vector<ShaderPreLoader::ShaderPermutation> ShaderPreLoader::canvas_shaders = Vector<ShaderPreLoader::ShaderPermutation>();
+Map<uint32_t, String> ShaderPreLoader::particle_shader_code = Map<uint32_t, String>();
+Vector<ShaderPreLoader::ShaderPermutation> ShaderPreLoader::particle_shaders = Vector<ShaderPreLoader::ShaderPermutation>();
+
+Error ShaderPreLoader::_load(VS::ShaderMode p_mode, const String &p_file_path) {
+	if (compiling) {
+		print_error("Can not load new shaderlists while already compiling. Please wait for the current compilation to finish!");
+		return Error::ERR_BUSY;
+	}
+
+	Map<uint32_t, String> *code_map;
+	Vector<ShaderPreLoader::ShaderPermutation> *permutations;
+	int *count;
+	if (p_mode == VS::ShaderMode::SHADER_SPATIAL) {
+		code_map = &spatial_shader_code;
+		permutations = &spatial_shaders;
+		count = &spatial_count;
+	} else if (p_mode == VS::ShaderMode::SHADER_CANVAS_ITEM) {
+		code_map = &canvas_shader_code;
+		permutations = &canvas_shaders;
+		count = &canvas_count;
+	} else if (p_mode == VS::ShaderMode::SHADER_PARTICLES) {
+		code_map = &particle_shader_code;
+		permutations = &particle_shaders;
+		count = &particle_count;
+	} else {
+		return Error::ERR_INVALID_PARAMETER;
+	}
+	progress = 0;
+	code_map->clear();
+	permutations->clear();
+
+	FileAccessCompressed *fa = memnew(FileAccessCompressed);
+	fa->configure("GCPF");
+	Error err = fa->_open(p_file_path, FileAccess::READ);
+	ERR_FAIL_COND_V_MSG(err, err, "ShaderPreLoader cannot open file '" + p_file_path + "'. Check user read permissions & the file location.");
+
+	//Load shader code
```

Comment:
```suggestion
	// Load shader code.
```
[Comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide) nitpick